io.openems.edge.batteryinverter.deye/src/io/openems/edge/deye/common/DeyeSunHybridImpl.java

```diff
@@ -0,0 +1,505 @@
+import io.openems.edge.ess.api.ManagedSymmetricEss;
+import io.openems.edge.ess.api.SymmetricEss;
+import io.openems.edge.ess.power.api.Constraint;
+import io.openems.edge.ess.power.api.Phase;
+import io.openems.edge.ess.power.api.Power;
+import io.openems.edge.ess.power.api.Pwr;
+import io.openems.edge.ess.power.api.Relationship;
+import io.openems.edge.timedata.api.Timedata;
+import io.openems.edge.timedata.api.TimedataProvider;
+import io.openems.edge.timedata.api.utils.CalculateEnergyFromPower;
+
+@Designate(ocd = Config.class, factory = true)
+@Component(//
+		name = "Deye.Sun.Hybrid - MaMoTec", //
+		immediate = true, //
+		configurationPolicy = ConfigurationPolicy.REQUIRE //
+)
+@EventTopics({ //
+		EdgeEventConstants.TOPIC_CYCLE_AFTER_PROCESS_IMAGE, //
+		EdgeEventConstants.TOPIC_CYCLE_BEFORE_CONTROLLERS //
+})
+public class DeyeSunHybridImpl extends AbstractOpenemsModbusComponent implements DeyeSunHybrid, ManagedSymmetricEss,
+		SymmetricEss, HybridEss, ModbusComponent, OpenemsComponent, EventHandler, ModbusSlave, TimedataProvider {
+
+	protected static final int MAX_APPARENT_POWER = 40000;
+
+	protected static final int NET_CAPACITY = 40000;
+
+	private static final int MIN_REACTIVE_POWER = -10000;
+
+	private static final int MAX_REACTIVE_POWER = 10000;
+
+	private final Logger log = LoggerFactory.getLogger(DeyeSunHybridImpl.class);
+
+	private final CalculateEnergyFromPower calculateAcChargeEnergy = new CalculateEnergyFromPower(this,
+			SymmetricEss.ChannelId.ACTIVE_CHARGE_ENERGY);
+
+	private final CalculateEnergyFromPower calculateAcDischargeEnergy = new CalculateEnergyFromPower(this,
+			SymmetricEss.ChannelId.ACTIVE_DISCHARGE_ENERGY);
+
+	private final CalculateEnergyFromPower calculateDcChargeEnergy = new CalculateEnergyFromPower(this,
+			HybridEss.ChannelId.DC_CHARGE_ENERGY);
+
+	private final CalculateEnergyFromPower calculateDcDischargeEnergy = new CalculateEnergyFromPower(this,
+			HybridEss.ChannelId.DC_DISCHARGE_ENERGY);
+
+	private final List<DeyeSunPv> chargers = new ArrayList<>();
+
+	private final SurplusFeedInHandler surplusFeedInHandler = new SurplusFeedInHandler(this);
+
+	@Reference
+	private ComponentManager componentManager;
+
+	@Reference
+	private Power power;
+
+	@Reference
+	private ConfigurationAdmin cm;
+
+	@Override
+	@Reference(policy = ReferencePolicy.STATIC, policyOption = ReferencePolicyOption.GREEDY, cardinality = ReferenceCardinality.MANDATORY)
+	protected void setModbus(BridgeModbus modbus) {
+		super.setModbus(modbus);
+	}
+
+	@Reference(policy = ReferencePolicy.DYNAMIC, policyOption = ReferencePolicyOption.GREEDY, cardinality = ReferenceCardinality.OPTIONAL)
+	private volatile Timedata timedata = null;
+
+	private Config config;
+
+	public DeyeSunHybridImpl() {
+		super(//
+				OpenemsComponent.ChannelId.values(), //
+				ModbusComponent.ChannelId.values(), //
+				SymmetricEss.ChannelId.values(), //
+				ManagedSymmetricEss.ChannelId.values(), //
+				HybridEss.ChannelId.values(), //
```

Comment:
Are all of these really needed ? Seems like most of them are Commercial specific...